task_1_2_2/src/main/java/ru/nsu/fit/tretyakov/Tree.java

```
@@ -199,81 +203,156 @@ public boolean containsAll(Collection<?> c) {
         boolean allHasBeenRetained = true;
-        boolean hasRoot = false;
-
-        for(Object o: c){
-            if (!contains(o)) allHasBeenRetained = false;
-            else{
-                if (o == null){
-                    throw new NullPointerException("Cannot retain null-element in collection");
-                }
-                if (!hasRoot){
-                    newTree = new Tree<>(1, (T)o);
-                    hasRoot = true;
-                }
-                else newTree.add((T)o);
+        TreeIterator<T> itr = this.iterator();
+
+        while (itr.hasNext()) {
+            var data = itr.peek();
+            if (!c.contains(data)) {
+                itr.remove();
+                this.size--;
+            } else {
+                allHasBeenRetained = false;
+                itr.next();
             }
         }
 
-        if (newTree != null){
-            root = newTree.root;
-            size = newTree.size();
-
-        }
         return allHasBeenRetained;
     }
 
+    /**
+     * Function that clears all nodes in the Tree.
+     */
     @Override
     public void clear() {
-        root = clearTraverse(root);
+        for (var elem : this) {
+            if (!elem.equals(root.data))
+                remove(elem);
+        }
+        remove(root.data);
         size = 0;
     }
 
-    private static class TreeIterator<T> implements Iterator<T>{
-        private List<Node<T>> treeNodes;
-        private Integer cursor;
-
-        private TreeIterator(Node<T> root) {
-            this.treeNodes = new ArrayList<>();
-            cursor = 0;
-            createArrayList(treeNodes,root);
+    /**
+     * This class is the iterator of the Tree.
+     * Iterators can iterate over the collections using BFS and check existing of next elements.
+     * They also can remove elements from the collection.
+     * @param <T> is the type of the stored objects.
+     */
+    public static class TreeIterator<T> implements Iterator<T> {
```

Не вижу поддержки ConcurrentModificationException, рассказывал на паре ,см. fail fast iterators